test/model/model.jl

```diff
@@ -297,248 +296,391 @@ function sample_gaussian_random_field!(field::AbstractMatrix{T},
+            model_data.model_params.nx, 
+            model_data.model_params.ny, 
+            model_data.model_params.n_state_var, 
+        )
+    )
+    if model_data.model_params.particle_initial_state == "true"
+        # Set true initial state
+        LLW2d.initheight!(
+            @view(state_fields[:, :, 1]), 
+            model_data.model_matrices, 
+            model_data.model_params.dx, 
+            model_data.model_params.dy,
+            model_data.model_params.source_size, 
+            model_data.model_params.peak_height, 
+            model_data.model_params.peak_position
+        )
+        state_fields[:, :, 2:3] .= 0
+    else
+        state_fields .= 0
     end
-
     # Add samples of the initial random field to all particles
-    add_random_field!(states.particles, field_buffer, initial_grf, rng, params.n_state_var, nprt_per_rank)
-
-end
-
-function set_stations!(stations::StationVectors, params::ModelParameters) where T
-
-    set_stations!(stations.ist,
-                  stations.jst,
-                  params.station_filename,
-                  params.station_distance_x,
-                  params.station_distance_y,
-                  params.station_boundary_x,
-                  params.station_boundary_y,
-                  params.dx,
-                  params.dy)
-
-end
-
-function set_stations!(ist::AbstractVector, jst::AbstractVector, filename::String, distance_x::T, distance_y::T, boundary_x::T, boundary_y::T, dx::T, dy::T) where T
-
-    if filename != ""
-        coords = readdlm(filename, ',', Float64, '\n'; comments=true, comment_char='#')
-        ist .= floor.(Int, coords[:,1] / dx)
-        jst .= floor.(Int, coords[:,2] / dy)
+    add_random_field!(
+        state_fields, 
+        view(model_data.field_buffer, :, :, 1, threadid()), 
+        model_data.initial_state_grf, 
+        rng,
+    )
+    return state
+end
+
+function get_station_grid_indices(params::ModelParameters)
+    if params.station_filename != ""
+        return get_station_grid_indices(
+            params.station_filename,
+            params.dx,
+            params.dy,
+        )
     else
-        LLW2d.set_stations!(ist,jst,distance_x,distance_y,boundary_x,boundary_y,dx,dy)
+        return get_station_grid_indices(
+            params.n_stations_x,
+            params.n_stations_y,
+            params.station_distance_x,
+            params.station_distance_y,
+            params.station_boundary_x,
+            params.station_boundary_y,
+            params.dx,
+            params.dy,
+        )
     end
 
 end
 
-struct ModelData{A,B,C,D,E,F,G,H}
-    model_params::A
-    states::B
-    observations::C
-    stations::D
-    field_buffer::E
-    background_grf::F
-    model_matrices::G
-    rng::H
+function get_station_grid_indices(
+    filename::String,
+    dx::T,
+    dy::T,
+) where T
+    coords = readdlm(filename, ',', Float64, '\n'; comments=true, comment_char='#')
+    return floor(Int, coords ./ [dx dy]) .+ 1
```

Comment:
This should be `floor.(Int, coords ./ [dx dy]) .+ 1`

Reply:
Thanks, just fixed